### PR TITLE
[WIP] Add plugins to Telegraf

### PR DESCRIPTION
**Pull request by philipnrmn**
## High-level description

This PR adds plugins to telegraf which gather metrics from Mesos about containers and from tasks running on containers via Mesos. It also adds a plugin which tags metrics with dcos primitives - name service, executor and task name. Finally it adds configuration for a prometheus plugin to expose these metrics. 

Finally, it moves dcos-metrics' prometheus producer to a different port ahead of removing it entirely (this work will come in a separate plugin).

## Corresponding DC/OS tickets (obligatory)

These DC/OS JIRA ticket(s) must be updated (ideally closed) in the moment this PR lands:

  - [DCOS_OSS-3714](https://jira.mesosphere.com/browse/DCOS_OSS-3714) - Adopt Telegraf as the DC/OS metrics pipeline
  - [DCOS_OSS-3716](https://jira.mesosphere.com/browse/DCOS_OSS-3716) - Write a Telegraf input plugin to get data for each container
  - [DCOS_OSS-3718](https://jira.mesosphere.com/browse/DCOS_OSS-3718) - Add plugin to receive dogstatsd metrics from the mesos module
  - [DCOS_OSS-3943](https://jira.mesosphere.com/browse/DCOS_OSS-3943) - Write a Telegraf processor plugin to add tags for each DC/OS-related metric

## Related tickets (optional)

Other tickets related to this change:

  - [DCOS_OSS-3719](https://jira.mesosphere.com/browse/DCOS_OSS-3719) Add an output plugin compatible with the metrics /v0 API 

## Checklist for all PRs

  - [x] Added a comprehensible changelog entry to `CHANGES.md` or explain why this is not a user-facing change:
  - [x] Included a test which will fail if code is reverted but test is not. If there is no test please explain here:
  - [x] Read the [DC/OS contributing guidelines](https://github.com/dcos/dcos/blob/master/contributing.md)
  - [x] Followed relevant code rules [Rules for Packages and Systemd](https://github.com/dcos/dcos/tree/master/docs)

## Checklist for component/package updates:

If you are changing components or packages in DC/OS (e.g. you are bumping the sha or ref of anything underneath `packages`), then in addition to the above please also include:

  - [ ] Change log from the last version integrated (this should be a link to commits for easy verification and review): [example](https://github.com/dcos/dcos-mesos-modules/compare/f6fa27d7c40f4207ba3bb2274e2cfe79b62a395a...6660b90fbbf69a15ef46d0184e36755881d6a5ae)
  - [ ] Test Results: [link to CI job test results for component]
  - [ ] Code Coverage (if available): [link to code coverage report]

### CODE REVIEW

**Comment by branden** [comment]
The cluster is failing to deploy because Mesos can't find the metrics isolator:

```
E0830 18:26:25.410120 28378 main.cpp:491] EXIT with status 1: Failed to create a containerizer: Could not create MesosContainerizer: Unknown or unsupported isolator 'com_mesosphere_MetricsIsolatorModule'
```

**Comment by philipnrmn** [comment]
@mesosphere-mergebot bump-ee

**Comment by philipnrmn** [comment]
@mesosphere-mergebot bump-ee

**Comment by philipnrmn** [comment]
@mesosphere-mergebot bump-ee

**Comment by branden** [comment]
@mesosphere-mergebot bump-ee

**Comment by philipnrmn** [comment]
@mesosphere-mergebot bump-ee

**Comment by klueska** [comment]
@mesosphere-mergebot bump-ee

**Comment by branden** [comment]
@mesosphere-mergebot bump-ee

**Comment by branden** [comment]
@mesosphere-mergebot bump-ee

**Comment by klueska** [comment]
@mesosphere-mergebot bump-ee

**Comment by klueska** [comment]
@mesosphere-mergebot bump-ee

**Comment by klueska** [comment]
@mesosphere-mergebot bump-ee

**Comment by klueska** [comment]
@mesosphere-mergebot label Ready For Review

**Comment by klueska** [comment]
@orsenthil I think this PR is as good as it's going to get at this point. The new service seems to fail on azure (probably because of some ubuntu 1.16 issues) and I will investigate them in the coming weeks while this is in soak.

**Comment by orsenthil** [comment]
The azure ARM failures can be ignored / overridden.

Enterprise side has a failure as noted in this comment: https://github.com/mesosphere/dcos-enterprise/pull/3347#issuecomment-418556472

**Comment by orsenthil** [comment]
@mesosphere-mergebot label ready for review

**Comment by orsenthil** [comment]
On the last test run on `teamcity/dcos/test/aws/onprem/static` suite

`test_applications.test_if_docker_app_can_be_deployed ` failed with an error.

```
E               AssertionError: Application deployment failed, reason: Failed to launch container: discarded; Abnormal executor termination: unknown container
```

I have re-triggered to determine if this a (new) flake.  The changes being introduced by the bump to mesos module are highly related (https://github.com/dcos/dcos/pull/3366/commits/d95d27ca5ea3ca7325e0e143b7bcae146b8bd906)

**Comment by branden** [comment]
For what it's worth, I'm unable to reproduce that failure locally with dcos-e2e:

```
branden@crateria-297 ~$ dcos-docker run "pytest -vvv -k test_applications"
============================================================ test session starts ============================================================
platform linux -- Python 3.6.3, pytest-3.1.0, py-1.4.31, pluggy-0.4.0 -- /opt/mesosphere/packages/python--2ba17816103010ce6dbbb2e263ade772ad0f73d9/bin/python3
cachedir: .cache
rootdir: /opt/mesosphere/packages/dcos-integration-test--691d4380a8869f83045232faf99073bfcbfb8692, inifile:
plugins: teamcity-messages-1.19, dcos-test-utils-0.1
collected 158 items

test_applications.py::test_if_marathon_app_can_be_deployed PASSED
test_applications.py::test_if_docker_app_can_be_deployed PASSED
test_applications.py::test_if_docker_app_can_be_deployed_windows SKIPPED
test_applications.py::test_if_ucr_app_can_be_deployed[Healthcheck.HTTP] PASSED
test_applications.py::test_if_ucr_app_can_be_deployed[Healthcheck.MESOS_HTTP] PASSED
test_applications.py::test_if_marathon_app_can_be_deployed_with_mesos_containerizer PASSED
test_applications.py::test_if_marathon_pods_can_be_deployed_with_mesos_containerizer PASSED
test_applications.py::test_octarine PASSED
```

**Comment by branden** [comment]
`test_applications.test_if_docker_app_can_be_deployed` passed this time! https://teamcity.mesosphere.io/viewLog.html?tab=buildLog&logTab=tree&filter=debug&expand=all&buildId=1193933&_focus=697

Maybe that failure was a flake?

**Comment by branden** [comment]
The failures on `teamcity/dcos/test/aws/cloudformation/simple` are all `test_networking.test_vip[Container_POD*`, and appear to be flaky, judging from their history.

The only failing test on `teamcity/dcos/test/aws/onprem/static` is `test_applications.test_if_ucr_app_can_be_deployed[Healthcheck_MESOS_HTTP]`. That test has failed on master a couple times in the last few days, so I assume that's a flake as well.

**Comment by klueska** [comment]
Azure issues to be followed up on in:
https://jira.mesosphere.com/browse/DCOS_OSS-4045

**Comment by branden** [comment]
@mesosphere-mergebot bump-ee

**Comment by branden** [comment]
@mesosphere-mergebot bump-ee

**Comment by branden** [comment]
Only a couple test failures on required checks this time.

[teamcity/dcos/test/aws/onprem/static](https://teamcity.mesosphere.io/viewLog.html?buildId=1194278&buildTypeId=DcOs_Open_Test_IntegrationTest_AwsOnpremWStaticBackend) has a single failed `test_networking.test_vip` test, with a history of recent failures on master. Looks flaky. Telegraf logs on this cluster don't show any problems

[teamcity/dcos/test/dcos-docker/static](https://teamcity.mesosphere.io/viewLog.html?buildId=1194256&buildTypeId=DcOs_Open_Test_DcosDocker_ProofOfConceptDcOsE2eDockerIntegrationTests) also has a single failed `test_networking.test_vip` test. This one doesn't have any recent failures on master, though it has failed on a few other PRs over the last week. I'm leaning toward this being a flake. This job doesn't provide logs, so I wasn't able to check telegraf's.

**Comment by klueska** [comment]
@jgehrcke How do you feel about this PR and the tests that are failing on it?

Test failure summaries:
https://github.com/dcos/dcos/pull/3366#issuecomment-418649170
https://github.com/dcos/dcos/pull/3366#issuecomment-418490167

**Comment by orsenthil** [comment]
@klueska 

1. test_vip failures can be considered as a flake, although the [frequency of failure](https://teamcity.mesosphere.io/viewType.html?buildTypeId=DcOs_Open_Test_DcosDocker_ProofOfConceptDcOsE2eDockerIntegrationTests&branch_DcOs_Open_Test_DcosDocker=pull%2F3366&tab=buildTypeStatusDiv) is much higher for the changes introduced by this PR.
2. [ACS Engine - Hybrid Cloud failure](https://teamcity.mesosphere.io/viewLog.html?buildId=1194354)  can be attributed to DCOS-41324 as we see the same [test failing on all PRs](https://teamcity.mesosphere.io/viewType.html?buildTypeId=DcOs_Open_Test_IntegrationTest_AcsEngineHybridCloud&tab=buildTypeHistoryList&branch_DcOs_Open_Test_IntegrationTest=__all_branches__) recently and on [master](https://teamcity.mesosphere.io/viewType.html?buildTypeId=DcOs_Open_Test_IntegrationTest_AcsEngineHybridCloud&branch_DcOs_Open_Test_IntegrationTest=%3Cdefault%3E&tab=buildTypeStatusDiv) due to unrelated change (supporting the notion that this is an external infrastructure, Azure's, issue)

3. The `dcos-checks-poststart.service` [failures](https://teamcity.mesosphere.io/viewLog.html?buildId=1194427) on Azure `teamcity/dcos/test/azure/arm` is new and a real bug. I see this is being tracked here - https://jira.mesosphere.com/browse/DCOS_OSS-4045  . Personally, I am okay with this being fixed out-of-band and overriding this failure.  We can determine if this bug is due to flaky azure infra or DC/OS service issue and probably fix this well before the beta release.

**Comment by jgehrcke** [comment]
@mesosphere-mergebot override-status teamcity/dcos/test/azure/arm https://jira.mesosphere.com/browse/DCOS_OSS-4045

**Comment by jgehrcke** [comment]
@mesosphere-mergebot override-status teamcity/dcos/test/azure/hybrid-cloud https://jira.mesosphere.com/browse/DCOS-41324

**Comment by jgehrcke** [comment]
@mesosphere-mergebot override-status teamcity/dcos/test/aws/onprem/static https://jira.mesosphere.com/browse/DCOS_OSS-2115

**Comment by klueska** [comment]
Thanks @orsenthil 

To clarify, I think merging even with the failures on Azure are justified because:

1) The are only about the service coming up on Ubuntu 1.16 (what we run on Azure)
2) This won't affect soak at all (we run on CentOS)
3) We will investigate and fix this issue in the next RI (https://jira.mesosphere.com/browse/DCOS_OSS-4045)

**Comment by orsenthil** [comment]
@mesosphere-mergebot override-status teamcity/dcos/test/dcos-docker/static https://jira.mesosphere.com/browse/DCOS_OSS-2115

File: CHANGES.md [hunk 1]
```diff
@@ -90,3 +90,5 @@ Format of the entries must be.
 * Updated REX-Ray version to 0.11.2 (DCOS_OSS-3597) [rexray v0.11.2](https://github.com/rexray/rexray/releases/tag/v0.11.2)
 
 * DC/OS can now be installed with SELinux in enforcing mode with the "targeted" policy loaded (DCOS-38953)
+
+* Replaced dcos-metrics with Telegraf (DCOS_OSS-3714)
```
Comment (jgehrcke) [inline CHANGES.md]: Note to self: we need a 'breaking changes' and a "what's new" section for a .0 release.

**Comment by jgehrcke** [comment]
I would feel most comfortable if @vespian and @gpaul had another look!

File: packages/telegraf/extra/dcos-telegraf.service [hunk 2]
```diff
@@ -7,12 +7,18 @@ StartLimitInterval=0
...
 PermissionsStartOnly=True
 
 EnvironmentFile=/opt/mesosphere/environment
 EnvironmentFile=/opt/mesosphere/etc/telegraf.env
 
 LimitNOFILE=16384
 
+ExecStartPre=/usr/bin/mkdir -p /run/dcos/telegraf
+ExecStartPre=/usr/bin/chmod 775 /run/dcos/telegraf
+ExecStartPre=/usr/bin/chown root:dcos_telegraf /run/dcos/telegraf
```
Comment (jgehrcke) [inline packages/telegraf/extra/dcos-telegraf.service]: CC @gpaul does this look good to you?

File: packages/telegraf/extra/dcos-telegraf.socket [hunk 3]
```diff
@@ -0,0 +1,7 @@
+[Unit]
+Description=Telegraf Socket: socket for Telegraf
+[Socket]
+ListenStream=/run/dcos/telegraf-dcos-metrics.sock
+SocketUser=root
+SocketGroup=dcos_adminrouter
+SocketMode=0660
```
Comment (jgehrcke) [inline packages/telegraf/extra/dcos-telegraf.socket]: CC @gpaul does this look good to you?

**Comment by urbanserj** [comment]
@philipnrmn could you please remove "[WIP]" label?

**Pull request merged**